packages/assets/src/index.ts

```diff
@@ -0,0 +1,589 @@
+    const chunks = Array.from<number[]>({ length: Math.ceil(this.length / this.chunkSize) });
+    await this.getChunks((index, chunk) => (chunks[index] = Array.from(chunk)));
+    return chunks.flat();
+  }
+
+  /**
+   * Write asset content to file (Node.js)
+   */
+  public async write(path: string): Promise<void> {
+    const fd = await new Promise<number>(async (resolve, reject) =>
+      fs.open(path, 'w', (err: any, fd: number) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+        resolve(fd);
+      }),
+    );
+    await this.getChunks(
+      (index, chunk) =>
+        new Promise<void>((resolve, reject) =>
+          fs.write(fd, chunk, 0, chunk.length, index * this.chunkSize, (err: any) => {
+            if (err) {
+              reject(err);
+              return;
+            }
+            resolve();
+          }),
+        ),
+    );
+  }
+
+  /**
+   * Get All chunks of asset through `onChunk` callback, can be used for a custom storage implementation
+   * @param onChunk Called on each received chunk
+   * @param sequential Chunks are received in sequential order when true or `concurrency` is `1` in config
+   */
+  public async getChunks(
+    onChunk: (index: number, chunk: Uint8Array) => void,
+    sequential?: boolean,
+  ) {
+    onChunk(0, this._content);
+    const chunkLimit = sequential ? limit(1) : this._limit;
+    await Promise.all(
+      Array.from({ length: Math.ceil(this.length / this.chunkSize) - 1 }).map((_, index) =>
+        chunkLimit(async () => {
+          const { content } = await this._actor.get_chunk({
+            key: this._key,
+            content_encoding: this.contentEncoding,
+            index: BigInt(index + 1),
+            sha256: this.sha256 ? [this.sha256] : [],
+          });
+          onChunk(index + 1, content);
+        }),
+      ),
+    );
+  }
+
+  /**
+   * Check if asset has been certified, which means that the content's hash is in the canister hash tree
+   */
+  public async isCertified(): Promise<boolean> {
+    // Below implementation is based on Internet Computer service worker
+    const agent = Actor.agentOf(this._actor) ?? getDefaultAgent();
+    const canisterId = Actor.canisterIdOf(this._actor);
+
+    if (!agent.rootKey) {
+      throw Error('Agent is missing root key');
+    }
+
+    const response = await this._limit(() =>
+      this._actor.http_request({
+        method: 'get',
+        url: this._key,
+        headers: [['Accept-Encoding', this._acceptEncodings.join(', ')]],
+        body: new Uint8Array(),
+      }),
+    );
+
+    let certificate: ArrayBuffer | undefined;
+    let tree: ArrayBuffer | undefined;
+    let encoding = '';
+    for (const [key, value] of response.headers) {
+      switch (key.trim().toLowerCase()) {
+        case 'ic-certificate':
+          {
+            const fields = value.split(/,/);
+            for (const f of fields) {
+              // @ts-ignore
```

Comment:
we prefer `as unknown as <expected-type>` when the TS engine can't pick up on an expected type